Add settings-based "global flags"

willbarton opened the pull request:
Here's support for what I'm calling "global flags" — flags that are defined in Django settings and apply to all Wagtail `Site`s that exist and override any `Site`-specific flags of the same name. "Global flags" are defined in settings as:

```python
FLAGS = {
    'MY_FLAG': True,
}
```

Per @chosak's suggestion this PR also adds support for FLAG_SOURCES:

```python
FLAG_SOURCES = (
    'my.module',
)
```

In my/module.py:

```python
MY_FLAG = True
```


As part of this change, I've updated [the README and restructured the "usage" section slightly](https://github.com/cfpb/wagtail-flags/blob/settings-flags/README.md#usage). I've also added a table to the top of the feature flags listing in the Wagtail admin to display any "global flags" in a read-only way:

![listing screenshot](https://github.com/cfpb/wagtail-flags/blob/settings-flags/screenshot_list.png?raw=true)

I have also (per regroup discussion) refactored the API a bit. *This is a breaking change and will necessitate a 2.0 release.*

`flag_enabled()` and `flag_disabled()` now take the flag name first and then an *optional* `HttpRequest` or `Site` object. If only the flag name is given, only global flags are checked. If the `HttpRequest` or `Site` are given, they'll check for any global flags first before checking the `FlagState` for the `Site`.

There is no change to the Django template API, because the Django template tags take the context object automatically.

There are a couple other minor niceties here:

- I added a check to the admin template to prompt the user to add a flag if none exist, instead of displaying an empty table, consistent with other Wagtail admin templates:

  ![image](https://cloud.githubusercontent.com/assets/10562538/24084716/e492920e-0cc5-11e7-9d66-397ad48b9352.png)

- Along the same lines, the global flags table won't appear in the admin if no global flags exist.

chosak commented:
This is a great idea.

What do you think about supporting multiple importable global flags, either as a replacement for or in addition to `settings.FLAGS`? I'm thinking individual apps might want to define their own flags, so you could do something like:

 ```py
 # settings.py
 FLAGS = {
     'foo': True,
     'bar': False,
     ...
 }
 
 FLAG_SOURCES = (
     'knowledgebase.flags',
     'v1.flags',
     ...
 )
 ``` 

 This is similar to how Django allows for import of template processors, middleware, etc. from named modules. It would also allow individual apps to define and manage their own app state without requiring a commit to the general `settings.py`.

Managing these efficiently might mean some kind of cached global set of flags which is generated once on first settings load (maybe during this app's `AppConfig.ready`?) and can then be efficiently iterated over in future. Potential complication would be how to handle multiply-defined flags.

willbarton commented:
@chosak I like that idea. Let me see what I can do.

willbarton commented:
I'm working on a bit of an API refactor that we discussed at regroup that adds what @chosak suggested and makes it possible to check for global flags without necessarily having a `Site` object.

willbarton commented:
@chosak @rosskarchner @richaagarwal @higs4281 I've refactored the flags API and added `FLAG_SOURCES` per @chosak's suggestion. The [README](https://github.com/cfpb/wagtail-flags/blob/settings-flags/README.md#usage) is up-to-date with these changes as is the PR OP.

higs4281 commented on README.md:
```diff
@@ -123,45 +178,43 @@ urlpatterns = [
+{% if my_flag %}
   <div class="m-global-banner">
-    I’m a beta banner.   
+    I’m a the result of a feature flag.   
```
"I'm a the ..."

higs4281 commented on README.md:
```diff
@@ -123,45 +178,43 @@ urlpatterns = [
+{% if flag_enabled('MY_FLAG', request) %}
   <div class="m-global-banner">
-    I’m a beta banner.   
+    I’m a the result of a feature flag.   
```
Same as above.

higs4281 commented on README.md:
````diff
@@ -33,43 +33,96 @@ pip install wagtail-flags
+
+```python
+FLAG_SOURCES = (
+    'my.module'
````
single-value tuple needs a comma

willbarton commented:
@higs4281 doh! Let me take a look.

willbarton commented:
@higs4281 want to give it a try now?

higs4281 commented:
👍 global and sourced flags working now. this is sweet.

chosak commented:
Looking into the future where flag logic has the potential to be a bit more complicated (say, enabled for a certain host or user), how would you feel about loosening the logic of `flag_state` and `flag_enabled` to be less specific? So instead of this:

```py
def flag_state(flag_name, request_or_site=None)
def flag_enabled(flag_name, request_or_site=None)
def flag_disabled(flag_name, request_or_site=None)
```

we'd have this:

```py
def flag_state(flag_name, **kwargs)
def flag_enabled(flag_name, **kwargs)
def flag_disabled(flag_name, **kwargs)
```

This would use the same logic inside of `flag_state` that defaulted to database lookup if the flag isn't in global settings, but instead of using a hardcoded `request_or_site` variable, it would look for a `request` or `site` keyword in the `kwargs`.

This would give us flexibility in the future if we wanted to expand `flag_enabled` logic to support something like:

```py
if flag_enabled('MY_FLAG', user):
    # do something
```

chosak commented on flags/settings.py:
```diff
@@ -0,0 +1,43 @@
+        source = import_module(source_str)
+        for flag in (f for f in dir(source)
+                     if f.isupper() and isinstance(getattr(source, f), bool)):
+            SOURCED_FLAGS[flag] = getattr(source, flag)
```
Should this raise an exception if a flag name is duplicated across sources? This seems like something that could be sneaky and hard to debug if it happened.

willbarton replied:
I will raise an exception here.

chosak commented on flags/settings.py:
```diff
@@ -0,0 +1,43 @@
+
+    # Global flags override everything. If there is a flag in a module defined
+    # in settings.FLAG_SOURCES and a flag defined in settings.FLAGS,
+    # settings.FLAGS wins.
```
Same comment as above; why not raise an exception here?

Another approach might be to have this package provide its own [system check](https://docs.djangoproject.com/en/1.8/topics/checks/#system-check-framework) that verifies that flag names aren't reused.

willbarton replied:
I think for now raising an exception makes sense. I'll do that now.

willbarton replied:
Edited the above comment: I think in this case raising an exception doesn't make sense — in this case a global flag could override a database flag. It might be worth adding a warning to the admin template when there are overlapping flags and logging a warning. But raising an exception here will cause the site to 500.

chosak replied:
By "will cause the site to 500", do you mean, will prevent Django from starting up? This doesn't seem so terrible to me, it's a misconfiguration (using a check instead would make it more of a warning than a failure).

Are database flags even in the mix at this part of the code?

willbarton replied:
I misunderstood my own code — yes an exception here would be thrown as Django starts up.

I'm fairly certain that raising an exception in `add_flags_from_sources` for duplicate global or sourced flags will be sufficient, and we won't have to use `setdefault` here anymore.

willbarton commented:
Based on some conversation today, and because this is already a breaking change, I want to use the opportunity to rework some of this before we merge. I might close this and open a new PR with these and other changes very soon.
